File: python/decorator.py
# Komponente
interface Coffee {
    cost(): number;
    description(): string;
}

# Konkrete Komponente
class SimpleCoffee implements Coffee {
    cost(): number {
        return 5;
    }

    description(): string {
        return "Simple Coffee";
    }
}

# Dekorator
abstract class CoffeeDecorator implements Coffee {
    protected coffee: Coffee;

    constructor(coffee: Coffee) {
        this.coffee = coffee;
    }

    cost(): number {
        return this.coffee.cost();
    }

    description(): string {
        return this.coffee.description();
    }
}

# Konkreter Dekorator
class MilkDecorator extends CoffeeDecorator {
    cost(): number {
        return this.coffee.cost() + 2;
    }

    description(): string {
        return this.coffee.description() + ", Milk";
    }
}

# Beispielanwendung
let coffee: Coffee = new SimpleCoffee();
console.log(coffee.description(), "Cost:", coffee.cost());

let milkCoffee: Coffee = new MilkDecorator(coffee);
console.log(milkCoffee.description(), "Cost:", milkCoffee.cost());

File: javascript/decorator.js
// Component interface
class Coffee {
    cost() {
      return 5;
    }
  
    description() {
      return 'Simple Coffee';
    }
  }
  
  // Decorator interface
  class CoffeeDecorator {
    constructor(coffee) {
      this.coffee = coffee;
    }
  
    cost() {
      return this.coffee.cost();
    }
  
    description() {
      return this.coffee.description();
    }
  }
  
  // Concrete Component
  class SimpleCoffee extends Coffee {}
  
  // Concrete Decorator
  class MilkDecorator extends CoffeeDecorator {
    cost() {
      return this.coffee.cost() + 2;
    }
  
    description() {
      return this.coffee.description() + ', Milk';
    }
  }
  
  // Example application
  const simpleCoffee = new SimpleCoffee();
  
  console.log(simpleCoffee.description(), 'Cost:', simpleCoffee.cost());
  
  const milkCoffee = new MilkDecorator(simpleCoffee);
  
  console.log(milkCoffee.description(), 'Cost:', milkCoffee.cost());
  